### DISCUSSION TITLE
추상화는 언제 해줘야할까요?

### DISCUSSION THREAD

**Discussion by moonheekim0118** [woowacourse-study/2022-code-review-study#34] [Lotto] [answered]
흔히들 반복이 보이면 추상화를 하라고 하잖아요, 그런데 반복이 보일 때 마다 추상화를 하는게 과연 옳은 방법일까요?
괜한 추상화 정도만 높여서 코드의 난이도를 높이는게 아닐지~ 걱정이되어요.
여러분은 어느때 추상화를 하시나요? 단순히 추상클래스 정의 뿐만 아니라 함수를 추상화 할 때에도 마찬가지입니다!
여러분만의 추상화 기준이 궁금해요

**Comment by wonsss** [comment]
일단 저는 사실 `추상화`가 무엇인지도 잘 몰라서 관련 개념을 찾아봤습니다. 

위키백과에서는 `컴퓨터과학에서 추상화(abstraction)는 복잡한 자료, 모듈, 시스템 등으로부터 핵심적인 개념 또는 기능을 간추려 내는 것을 말한다.`라고 정의하고 있네요.

즉, 추상화는 복잡한 것들을 목적에 맞게 단순화하는 것이네요.

추상화를 통해 코드의 재사용성, 가독성, 생산성을 향상시키고, 에러를 감소할 수 있다고 하던데, 
결국 추상화의 효용성은 기능 개발 후 `"유지 보수, 버그 해결" ` 때 나온다고 할 수 있을 것 같아요.

따라서, 지금 만든 코드가 그저 잘 돌아가는 것으로 만족하지 않고,
본인이 1개월 또는 1년 후에, 아니면 다른 사람이 보게 되었을 때,
세부적인 것들을 일일이 보지 않고도 잘 그려진 추상화(그림)를 보는 것처럼
`한 눈에 이해되는 코드`로 만드는 것이 추상화의 목적이 아닐까요?

이런 추상화의 목적을 생각하면서, 설계를 하고 이후에 다시 리팩토링을 하면서 
`같은 `특성``을 확인하게 되었고 이를 다른 곳에서 `재사용`할 수 있다고 판단될 때, 이를 추출하여 추상화를 해야할 것 같아요.

그렇지 않고 `미활용된 추상화 설계`는 `YAGNI` 원칙에 위배될 것 같으니, 설계에 기여하지 않는 과도한 추상화는 지양해야 할 것 같습니다.

별론으로서, 함수의 이름, 클래스의 이름, 파일의 이름, 디렉토리의 이름도 추상화의 일부라고 생각돼요.
디렉토리의 내용물을 살펴보지 않더라도 추상화가 잘된 디렉토리명만 보더라도 디렉토리가 하는 일을 무엇인지 알 수 있으니까요.

(방금 검색하면서 머릿속으로 정리한 내용이라, 틀릴 수 있습니다. 비판적으로 읽어주세요)

**Comment by byhhh2** [comment, marked as answer]
이 질문을 읽으니깐 저의 지식 자체가 흔들리고 있습니다..

추상화란 무엇?

1. 알 필요없는 복잡한 메서드를 가리는 것 
2. `놀다()`, `자다()` 등을 `하다()`로 추상화하는 것 - 여기서는 상속을 사용하겠죠?

<br />

호프의 질문은 2번이 맞을까요?! 2번이라면 저의 기준은 `도메인`일 것 같아요.
도메인이 다르지만 반복적인 행위가 있다면 추상화를 생각해볼 것 같습니다.

**Reply by moonheekim0118** [reply]
2번 맞습니다 '도메인은 다르지만 반복적인 행위가 있다' 좋은 의견입니다 👍
